lib/vector-core/src/transform/mod.rs

```diff
@@ -182,3 +142,164 @@ pub trait TaskTransform: Send {
     where
         Self: 'static;
 }
+
+/// Broader than the simple [`FunctionTransform`], this trait allows transforms to write to
+/// multiple outputs. Those outputs must be known in advanced and returned via
+/// `TransformConfig::named_outputs`. Attempting to send to any named output not registered in
+/// advance is considered a bug and will cause a panic.
+pub trait SyncTransform: Send + dyn_clone::DynClone + Sync {
+    fn transform(&mut self, event: Event, output: &mut TransformOutputsBuf);
+}
+
+dyn_clone::clone_trait_object!(SyncTransform);
+
+impl<T> SyncTransform for T
+where
+    T: FunctionTransform,
+{
+    fn transform(&mut self, event: Event, output: &mut TransformOutputsBuf) {
+        FunctionTransform::transform(self, &mut output.primary_buffer, event);
+    }
+}
+
+// TODO: this is a bit ugly when we already have the above impl
+impl SyncTransform for Box<dyn FunctionTransform> {
+    fn transform(&mut self, event: Event, output: &mut TransformOutputsBuf) {
+        FunctionTransform::transform(self.as_mut(), &mut output.primary_buffer, event);
+    }
+}
+
+pub struct TransformOutputs {
+    primary_output: Fanout,
+    named_outputs: HashMap<String, Fanout>,
```

Comment:
You're likely correct about the performance, but this isn't quite used the way you describe. Transforms write directly into a `TransformOutputsBuf` (which has the same structure, but `Vec<Event>` instead of `Fanout`), and this struct is than used to forward a whole batch of outputs downstream. So it seems unlikely to me that a hashmap access here is going to be meaningful in the scheme of things, but it's possible that the same criticism could apply more to `TransformOutputsBuf`. 

Your mention of indexing gave me the idea that we could use something similar to the `slotmap` crate, where inserting returns a unique key type. Then we could somehow provide those keys back to the transform to ensure via typechecking that transforms can't write to unknown keys, as well as gain the increased performance of direct indexing vs hashing. That'd require a significantly large change, so I'm inclined to think on it a bit more and defer to a later PR.

Reply:
The idea you describe is along the same lines as I was thinking. I am fine to leave that for follow-up work.